wagtail_localize/modeladmin/views.py

```diff
@@ -0,0 +1,164 @@
+from wagtail_localize.views import edit_translation
+
+
+if WAGTAIL_VERSION >= (2, 15):
+    from wagtail.contrib.modeladmin.views import HistoryView
+else:
+    HistoryView = object
+
+
+class TranslatableViewMixin:
+    def __init__(self, *args, **kwargs):
+        self.locale = None
+        super().__init__(*args, **kwargs)
+
+        if not issubclass(self.model, TranslatableMixin):
+            raise ImproperlyConfigured(
+                f"Model `{self.model}` used in translatable view `{self.__class__}` "
+                f"must subclass the `{TranslatableMixin}` class."
+            )
+
+    @method_decorator(login_required)
+    def dispatch(self, request, *args, **kwargs):
+        if getattr(self, "instance", None):
+            self.locale = self.instance.locale
+        if "locale" in request.GET:
+            self.locale = get_object_or_404(Locale, language_code=request.GET["locale"])
+        if not self.locale:
+            self.locale = Locale.get_active()
+        return super().dispatch(request, *args, **kwargs)
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+        context["locale"] = self.locale
+        context["translations"] = []
+        context["wagtail_version"] = get_main_version()
+        return context
+
+
+class TranslatableIndexView(TranslatableViewMixin, IndexView):
+    def get_filters(self, request):
+        filters = super().get_filters(request)
+        filters[2]["locale"] = self.locale.id
+        return filters
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+        context["translations"] = [
+            {
+                "locale": locale,
+                "url": self.index_url + "?locale=" + locale.language_code,
+            }
+            for locale in Locale.objects.exclude(id=self.locale.id)
+        ]
+        return context
+
+
+class TranslatableCreateView(TranslatableViewMixin, CreateView):
+    def get_form_kwargs(self):
+        kwargs = super().get_form_kwargs()
+        kwargs["instance"].locale = self.locale
+        return kwargs
+
+    def get_success_url(self):
+        return self.index_url + "?locale=" + self.locale.language_code
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+        context["translations"] = [
+            {
+                "locale": locale,
+                "url": self.create_url + "?locale=" + locale.language_code,
+            }
+            for locale in Locale.objects.exclude(id=self.locale.id)
+        ]
+        return context
+
+
+class TranslatableEditView(TranslatableViewMixin, EditView):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self.translation = Translation.objects.filter(
+            source__object_id=self.instance.translation_key,
+            target_locale_id=self.instance.locale_id,
+        ).first()
```

Comment:
nitpick: this does need a comment as to why `.first()` for future reference

Reply:
Added the extra comment.